libgadget/treewalk.c

```diff
@@ -281,47 +275,74 @@ cmpint(const void *a, const void *b)
 #endif
 
 static void
-treewalk_build_queue(TreeWalk * tw, int * active_set, int size) {
+treewalk_init_evaluated(int * active_set, int size)
+{
+    #pragma omp parallel for
+    for(int i=0; i < size; i++)
+    {
+        const int p_i = active_set ? active_set[i] : i;
+        P[p_i].Evaluated = 0;
+    }
+}
+
+static void
+treewalk_build_queue(TreeWalk * tw, int * active_set, const int size, int may_have_garbage) {
+    int i;
+
+    if(!tw->haswork && !may_have_garbage)
+    {
+        tw->WorkSetSize = size;
+        tw->WorkSet = active_set;
+        tw->work_set_allocated = 0;
+        return;
+    }
+
+    tw->WorkSet = mymalloc("ActiveQueue", size * sizeof(int));
+    tw->work_set_allocated = 1;
```

Comment:
Why not checking for NULL instead of a new variable?

Reply:
I see. It can be a stolen array.

Reply:
Use 'workset_stolen_from_active_set' and reverse the logic?

Reply:
Done